2-Draft-Accepted/RFCxxxx-Secrets-Management.md

```diff
@@ -0,0 +1,150 @@
+Best practice is to not hard code any secrets in scripts, but currently
+this requires custom code on different platforms to handle this securely.
+
+This RFC proposes new cmdlets to make this simpler and secure.
+
+## Motivation
+
+> As a PowerShell script author,
+> I can securely use multiple secrets,
+> so that I can automate complex orchestration across multiple remote resources.
+
+## High Level Design
+
+Secrets are stored securely in a local vault.
+The local vault is expected to only allow access to the user who owns that
+vault.
+Secrets required to access remote vaults are stored in the local vault and used by the secrets management
+cmdlets to retrieve remote secrets.
+
+`User Context` --> `Local Vault` --> `SecretsVaultExtension` --> `Remote Vault`
+
+## User Experience
+
+Registering and using remote secrets:
+
+```powershell
+# Install the Az.KeyVault module to be used in this example.
+# This module doesn't conform to the extensions model so would need to be
+# updated to actually work.
+Install-Module Az.KeyVault
+
+# In this example, we explicitly register this extension
+Register-SecretsVaultExtension -Name Azure -Cmdlet Get-AzKeyVaultSecret `
+  -Module Az.KeyVault -Version 1.0.0
+
+# When using remote vaults, expectation is that secrets are
+# stored using their tooling, so this module is only for retrieving secrets
+# from remote vaults
+$azDevOpsToken = Get-Secret -Name AzDevOps
+Invoke-RestMethod https://dev.azure.com/… -Credential $azDevOpsToken -Authentication Basic
+```
+
+Registering and using local secret:
+
+```powershell
+# For local vault, we can register custom secrets
+# In this example, we store a PSCredential object
+Register-Secret -Secret $cred -Name MyCreds
+New-PSSession -ComputerName myServer -Credential (Get-Secret -Name MyCreds)
+```
+
+## Specification
+
+### Secrets Vault
+
+There are two types of vaults: local and remote.
+A local vault, by default, is already created and named `Default`.
+`Default` on Windows is [CredMan](https://docs.microsoft.com/en-us/windows/desktop/SecAuthN/credentials-management).
+`Default` on macOS is [KeyChain](https://developer.apple.com/documentation/security/keychain_services).
+
+>[!NOTE]
+>KeyChain support is unlikely to be available in the first release of this feature.
+
+`Default` on Linux is [Hashicorp Vault](https://www.vaultproject.io/).
+
+>[!NOTE]
+>Linux has many options for local credential management, additional support may be
+>added in the future or via community provided extensions.
+
+### Credential Vault Extensions
+
+Vault extensions whether they are local or remote are provided as modules.
+
+Extensions are modules that provide a `Get-Secret` cmdlet that accepts a
+`-Credential` or `-AccessToken` parameter for authentication.
+Expectation is that each module will have their own cmdlet and not literally
+named `Get-Secret`.
+
+Extensions register themselves by calling `Register-SecretsVaultExtension`
```

Comment:
I am worried about this.  A malicious actor could register a secrets vault extension, or simply drop a malicious module that implements a secrets extension.  Then it could harvest secrets via the `Register-Secret` cmdlet.  Or it could DOS by returning garbage for retrieved secrets.

Reply:
As discussed, there's not much we can do if the hacker has control of the system and is able to run as that user context.  Rotating secrets on the target side is the best practice to mitigate this.